### PR TITLE
[complex] enable complex autograd cumsum

### LINKED ISSUE #53182: Add complex autograd support for torch.cumsum
```py
In [9]: x=torch.randn(2, 3, dtype=torch.cdouble, requires_grad=True)

In [10]: torch.cumsum(x, dim=0)
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-11-9aad052c3955> in <module>
----> 1 torch.cumsum(x, dim=0)

RuntimeError: cumsum does not support automatic differentiation for outputs with complex dtype.
```

cc @ezyang @anjali411 @dylanbespalko @mruberry

### PR DESCRIPTION
**Pull request by kshitij12345**
Fixes #53182

Turns out that there is no need to update formula :)

### CODE REVIEW

**Comment by kshitij12345** [comment]
**Note**: Few failures are real. Will fix and ping.

File: torch/testing/_internal/common_methods_invocations.py [hunk 1]
```diff
@@ -1678,6 +1694,31 @@ def _make_tensor_helper(shape, low=None, high=None):
...
+               # torch.int8
+               SkipInfo('TestCommon', 'test_variant_consistency_eager',
+                        dtypes=[torch.uint8, torch.int8, torch.int16, torch.int32]),
+               # >>> t = torch.tensor([False])
+               # >>> t.cumsum_(0) # Error "cumsum_out_cpu" not implemented for 'Bool'
+               # >>> t.cuda().cumsum_(0) # Error "cumsum_cuda" not implemented for 'Bool'
+               # >>> t = torch.tensor(False)
+               # >>> t.cumsum_(0) # Error "cumsum_out_cpu" not implemented for 'Bool'
+               # >>> t.cuda().cumsum_(0) # Does not fail!
+               # tensor(False, device='cuda:0')
```
Comment (kshitij12345) [inline torch/testing/_internal/common_methods_invocations.py:1717]: These lines allow for this behaviour

https://github.com/pytorch/pytorch/blob/b864457743d7fd0f7dc70386dd10eb3f31ba6f22/aten/src/ATen/native/cuda/ScanKernels.cu#L594-L601

File: torch/testing/_internal/common_methods_invocations.py [hunk 2]
```diff
@@ -1678,6 +1694,31 @@ def _make_tensor_helper(shape, low=None, high=None):
                        SkipInfo('TestCommon', 'test_variant_consistency_jit',
                                 device_type='cuda', dtypes=[torch.float16]),
                    )),
+    OpInfo('cumsum',
+           dtypesIfCPU=all_types_and_complex_and(torch.bool),
+           dtypesIfCUDA=all_types_and_complex_and(torch.bool, torch.half),
+           skips=(
+               # For integer inputs,
```
Comment (anjali411) [inline torch/testing/_internal/common_methods_invocations.py:1701]: Hey @kshitij12345 good catch! could you file an issue for this? I am not sure if this is the desired behavior. But even so, it's not documented so we could repurpose the issue to either fix this behavior or update the doc!

**Comment by kshitij12345** [comment]
@anjali411 You can have a look whenever free. 

I think we should enable bool support for inplace on both CPU and CUDA as method and functional variants allows bool.
```python
>>> t = torch.tensor(False)
>>> t.cumsum(0)
tensor(0)
>>> t.cuda().cumsum(0)
tensor(0, device='cuda:0')
```

Let me know what you feel about the same and should we do it in this PR.

Thanks!

File: torch/testing/_internal/common_methods_invocations.py (hunk 2 above)
Comment (kshitij12345) [reply torch/testing/_internal/common_methods_invocations.py:1701]: Will open an issue. Thanks!

Comment (anjali411) [reply torch/testing/_internal/common_methods_invocations.py:1701]: thanks and could you also link that issue in a comment above the skip?

File: torch/testing/_internal/common_methods_invocations.py [hunk 3]
```diff
@@ -1678,6 +1694,31 @@ def _make_tensor_helper(shape, low=None, high=None):
...
+               # always promotes it to torch.long.
+               # >>> t = torch.randint(2, 10, (3, 2), dtype=torch.int8)
+               # >>> t.cumsum(0).dtype
+               # torch.int64
+               # >>> t.cumsum_(0).dtype
+               # torch.int8
+               SkipInfo('TestCommon', 'test_variant_consistency_eager',
+                        dtypes=[torch.uint8, torch.int8, torch.int16, torch.int32]),
+               # >>> t = torch.tensor([False])
+               # >>> t.cumsum_(0) # Error "cumsum_out_cpu" not implemented for 'Bool'
```
Comment (anjali411) [inline torch/testing/_internal/common_methods_invocations.py:1712]: Let's file a separate issue requesting bool support for in-place operation on both CPU and CUDA.

**Comment by anjali411** [comment]
> I think we should enable bool support for in-place on both CPU and CUDA as method and functional variants allows bool.

I imagine the current behavior was set on purpose since if the tensor is 0 dim, it only has one element and then it's cumulative sum would either be 0 or 1 which is clearly mappable to True/False, otherwise the sum could be > 1 but that does still get casted to True in most(all?) languages to True. so yeah I think we should support in-place version of cumsum for bool. could you file an issue and link it in a comment above the skip?

> Let me know what you feel about the same and should we do it in this PR.

Let's add that in a separate PR!

File: torch/testing/_internal/common_methods_invocations.py [hunk 4]
```diff
@@ -1362,6 +1362,22 @@ def _make_tensor_helper(shape, low=None, high=None):
...
 
+
+def sample_inputs_cumsum(op_info, device, dtype, requires_grad):
+    def _make_tensor_helper(shape, low=None, high=None):
+        return make_tensor(shape, device, dtype, low=low, high=high, requires_grad=requires_grad)
+
+    samples = (
+        SampleInput((_make_tensor_helper((S, S, S)), 0)),
+        SampleInput((_make_tensor_helper((S, S, S)), 1)),
+        # FIXME : inplace grad fails for the following sample due to mismatch in
```
Comment (anjali411) [inline torch/testing/_internal/common_methods_invocations.py:1373]: does it only fail for complex inputs or both real and complex inputs?

Comment (kshitij12345) [reply torch/testing/_internal/common_methods_invocations.py:1373]: Changing it to
```
kwarg_dtype = torch.cdouble if dtype.is_complex else torch.double
SampleInput((_make_tensor_helper((S, S, S)), 1), kwargs={'dtype': kwarg_dtype}),
```

works as we do gradcheck with torch.double and torch.cdouble but 
`test_variant_consistency_eager_cumsum` fails for `float16, float32, cfloat` with type mismatch in the inplace variant check.

**Review by anjali411 — APPROVED** [review]
LGTM. thanks @kshitij12345 !

**Comment by anjali411** [comment]
@kshitij12345 could you please rebase this PR?

**Pull request closed without merging**